File: src/main/java/quizapi/service/UserService.java
package quizapi.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import quizapi.exception.ResourceNotFoundException;
import quizapi.model.User;
import quizapi.repository.UserRepository;

@Service
public class UserService {

    @Autowired
    private UserRepository userRepository;

    public User findById(Long id) {
        User user = userRepository
                .findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("User", "id", id));
        return user;
    }

    public User findOrCreateByUsername(String username) {
        User user;
        if (userRepository.existsByUsername(username)) {
            user = userRepository
                    .findByUsername(username)
                    .orElseThrow(() -> new ResourceNotFoundException("User", "username", username));
        } else {
            user = userRepository.save(new User(username));
        }
        return user;
    }

    public User findByUsername(String username) {
        return userRepository
                .findByUsername(username)
                .orElseThrow(() -> new ResourceNotFoundException("User", "username", username));
    }
}

File: src/main/kotlin/com/shkila/quizapi/service/UserService.kt
package com.shkila.quizapi.service

import com.shkila.quizapi.exception.ResourceNotFoundException
import com.shkila.quizapi.model.User
import com.shkila.quizapi.repository.UserRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
class UserService {

    @Autowired
    lateinit var userRepository: UserRepository

    fun findOrCreateByUsername(username: String): User {
        return if (userRepository.existsByUsername(username) == true) {
            userRepository
                    .findByUsername(username)
                    .orElseThrow { ResourceNotFoundException("User", "username", username) }
        } else {
            userRepository.save(User(username = username))
        }
    }

    fun findByUsername(username: String): User {
        return userRepository
                .findByUsername(username)
                .orElseThrow { ResourceNotFoundException("User", "username", username) }
    }
}
